Split webserver from node

Clintonio opened the pull request:
The web server is now split from the node at the process level. Changes;

* the node and web server run in a separate process
* the web server is not started by default with the driver
* the web server is only started with Cordformation built nodes if there is a web address configuration present
* starting the web server still uses the main corda.jar. This PR got too large to consider splitting this further.
* I am unsure on the JMX situation. If it is broken I'd rather re-add it after this PR to avoid any conflict.
* Cordapp template will get an update when this PR is approved.
* Requires new Cordformation to be published before merging.

PS. I'll squash this down later. Please ignore individual commits.

Clintonio commented:
Looks like I need a rebase.

mnesbit commented on node/src/main/kotlin/net/corda/node/webserver/WebServer.kt:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    private fun connectLocalRpcWithRetries(retries: Int): CordaRPCOps {
+        for(i in 0..retries - 1) {
```
I think finite retries is going to cause pain when we are debugging remotely and trying things out.

Clintonio replied:
Changed. I am not going with infinite retries. It will be less flaky and allows webservers to come up any time.

exFalso commented on node/src/main/kotlin/net/corda/node/driver/Driver.kt:
```diff
@@ -318,25 +327,20 @@ open class DriverDSL(
-            return om.readValue(conn.inputStream, NodeInfo::class.java)
+    private fun queryNodeInfo(nodeAddress: HostAndPort, sslConfig: NodeSSLConfiguration): NodeInfo? {
+        var retries = 0
+        while (retries < 5) try {
```
I think we decided to avoid timeouts like this in the code, as it may and will cause flakiness in e.g. integ tests. Timeouts should be handled at the topmost level (e.g. CI)

Clintonio replied:
What do you suggest then? The requirement of this function is;

1. It can verify the node is up
2. If the node isn't up it fails

But since this function is run _almost immediately_ after the node processes are kicked off this function can never succeed on  (1) because in normal usage (1) will never be true at the time this function is called. 

Since the driver is used in automated environments we can't rely on user input, so all we have is timeouts/waiting. The previous solution had an implicit timeout too on the HTTP request, which usually resolved in time (60s or so). We can either poll, like this, with retries, or we can have a long wait up front. If you know another solution though please do say.

exFalso commented on node/src/main/kotlin/net/corda/node/driver/Driver.kt:
```diff
@@ -413,12 +417,45 @@ open class DriverDSL(
+        val client = OkHttpClient.Builder().connectTimeout(5, TimeUnit.SECONDS).readTimeout(60, TimeUnit.SECONDS).build()
+        val retries = 50
+
+        for (i in 0..retries) {
```
Ditto

exFalso commented on node/src/main/kotlin/net/corda/node/utilities/JsonSupport.kt:
```diff
@@ -108,9 +129,10 @@ object JsonSupport {
+            val mapper = parser.codec as PartyObjectMapper
             // TODO this needs to use some industry identifier(s) not just these human readable names
-            return mapper.identities.partyFromName(parser.text) ?: throw JsonParseException(parser, "Could not find a Party with name: ${parser.text}")
+            return mapper.partyFromName(parser.text) ?: throw JsonParseException(parser, "Could not find a Party with name ${parser.text}")
```
This is... suboptimal. We are making an RPC request for each Party field! Why don't we just serialise the whole thing?

Clintonio replied:
I didn't write that and would prefer to not change it as a part of this PR. There's too many rabbit holes already.

exFalso replied:
Hmm can we trace back why we need this? This is the only parser that needs the servicehub/rpc stuff, and it doesn't seem necessary to me at all, it makes parsing side-effecting and non-deterministic, and it's inefficient

exFalso replied:
ping

mikehearn replied:
Can you add a TODO to revisit this please. The JSON support is something that I guess should move into the client library as part of the split anyway.

exFalso commented on node/src/main/kotlin/net/corda/node/webserver/WebServer.kt:
```diff
@@ -0,0 +1,183 @@
+class WebServer(val config: FullNodeConfiguration) {
+    private companion object {
+        val log = loggerFor<WebServer>()
+        val maxRetries = 60 // TODO: Make configurable
```
Same, we shouldn't have retries at this level

Clintonio replied:
This one is fixed. Infinite retries is fine at the process level. You can assume if you've started the webserver that you're expecting a node to appear at some point. If it doesn't then it's down to the user or CI to figure that out.

exFalso commented on node/src/main/kotlin/net/corda/node/webserver/WebServer.kt:
```diff
@@ -0,0 +1,183 @@
+
+    fun run() {
+        while(server.isRunning) {
+            Thread.sleep(100) // TODO: Redesign
```
I'm sure there are better ways of doing this...

Clintonio replied:
Indeed, hence the TODO. I could do it as a part of this PR, or I can do it in the next one.

mikehearn replied:
As the eventual goal is for the web apps to be entirely "normal" (in java world, means WARs) it's not worth polishing this code too much I guess.

mikehearn commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -82,6 +82,11 @@ interface CordaRPCOps : RPCOps {
     fun nodeIdentity(): NodeInfo
 
+    /**
+     * Returns true if node is up and ready
```
This is vague. What does "ready" mean. What does this returning true mean beyond "RPC is available"? When would you use it and why? A good api doc explains why an API exists:

https://docs.corda.net/codestyle.html#comments

Clintonio replied:
I copied this description from the old API endpoint. I will improve it though. I was aiming not to rabbit hole too much with this PR.

Clintonio replied:
I'll remove it from the RPC interface. It's no longer useful after further inspection. I'll keep it on the web so that there is an easy place to poll for a 200 response if you want to check the webserver is up. RPC already has functionality or TODOs surrounding that.

mikehearn commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -103,6 +108,11 @@ interface CordaRPCOps : RPCOps {
     /**
+     * Uploads a file by data type to the node and returns a meaningful string to the user
+     */
+    fun uploadFile(dataType: String, name: String?, file: InputStream): String
```
What does "meaningful string" mean? Is this suitable for placing in a user interface? If so, is it localised? Is the string unique? What does "data type" mean? MIME type? What happens to the uploaded file? Where does it go? Is there a way to delete the file? Size limits?

Enquiring minds want to know!

(btw, when a code review has questions like this, it is more like "the code should explain this as other programmers may want to know", not an invitation for answers on the code review thread)

Clintonio replied:
I copied that description from the original API endpoint.

Clintonio replied:
[This](https://github.com/corda/corda/blob/master/node/src/main/kotlin/net/corda/node/servlets/DataUploadServlet.kt) ported to the RPC interface. It was _very_ built into the node before. 

Remember that the node and web can be on different machines so the webserver can't hold any state, like file uploads, that the node needs, like the Oracle in the IRS demo.

mikehearn replied:
Yes, but that's a good opportunity to either improve the docs, or decide this isn't going to be actual API and delete the docs and mark it as deprecated (the actual call sites can have the warning suppressed).

mikehearn commented on core/src/main/kotlin/net/corda/core/node/services/Services.kt:
```diff
@@ -232,6 +244,9 @@ interface StorageService {
     /** Provides access to storage of arbitrary JAR files (which may contain only data, no code). */
     val attachments: AttachmentStorage
 
+    /** Provides file uploads of arbitrary files to services **/
```
Ah ha. So this is starting to sound like a way to insert data into an oracle?

Is this a workaround for inability of plugins to extend the RPC interface?

mikehearn replied:
Let's remove the API doc for this and mark it as deprecated, to indicate that this API will probably go away in future. I think for things like adding data to oracles, the right path is to improve the RPC framework rather than have a generic file upload mechanism. For now it's OK to unblock this PR.

Clintonio replied:
I have added a Deprecated annotation to both the service and RPC endpoint. I do agree that this should be removed, it doesn't seem secure.

mikehearn commented on docs/source/corda-plugins.rst:
```diff
@@ -34,21 +34,16 @@ of the node internal subsystems.
-    ``servicePlugins`` the ``webApis`` cannot register new protocols, or 
-    initiate threads. (N.B. The intent is to move the Web support into a 
-    separate helper process using the RPC mechanism to control access.) 
+    classes. These classes will be constructed by the bundled web server
```
???

I thought the point of this work was that there'd be no special API or support for web apps anymore? They'd just be "normal" java style web apps with no special support from Corda beyond the RPC client library?

Clintonio replied:
That will be split in a future PR. I just kept the docs up to date with this intermediate step.

mikehearn commented on docs/source/corda-plugins.rst:
```diff
@@ -34,21 +34,16 @@ of the node internal subsystems.
     paths and allows simple web demos to be distributed within the CorDapp 
-    jars. (N.B. The intent is to move the Web support into a separate helper 
-    process using the RPC mechanism to control access.) 
+    jars. These static serving directories will not be available if the
```
Ditto - why is this here? I thought we were trying to split the web stuff out of the node. Not just make it optional.

Clintonio replied:
Same as above.

mikehearn commented on docs/source/creating-a-cordapp.rst:
```diff
@@ -12,10 +12,10 @@ App plugins
 2. Service plugins: Register your services (see below).
-3. Web APIs: You may register your own endpoints under /api/ of the built-in web server.
-4. Static web endpoints: You may register your own static serving directories for serving web content.
+3. Web APIs: You may register your own endpoints under /api/ of the bundled web server.
```
???? What's the difference between "bundled" and "built in"

Clintonio replied:
Bundle - a collection of things (in this case distinct things). Built-in - it exists as a part of the other thing. I picked this word pretty carefully to convey there is now a webserver _and_ node whereas before it was webserver _in_ node.

Clintonio commented:
@mikehearn what did you expect it to be? This PR isn't the final piece, but it's a stepping stone. There is a lot of testing involved in this change and I split the server out first at the process level (while keeping it in the same module, same interfaces, etc) first since, as you can see, it's a fairly large change. Testing this took a significant amount of time.

Before I continue with this PR can we ensure we're on the same page on where this PR, and any further work from it, is expected to be. 

Also, as for that arbitrary attachment uploading; I had to keep that to keep the IRS sample working (Oracle rate upload). If there is a way to arbitrarily extend the RPC interface I can do that in the IRS sample, otherwise I'd like to keep that change to another PR to avoid making this one even larger.

Clintonio commented:
This PR isn't the final piece, but it's a stepping stone. There is a lot of testing involved in this change and I split the server out first at the process level (while keeping it in the same module, same interfaces, etc) first since, as you can see, it's a fairly large change. Testing this took a significant amount of time. 

So the current state is; the node itself has no web support at the process level. The node module contains the web code, this can be moved out. It was not easy at the start of the PR so I was hoping to keep code shovelling like that for follow on work. The node main method can also start the webserver but it is behind a flag. This is again, something I want to change when I split it into a separate module. The capsule contains the webserver. This I was unsure about, but it's not too much work to make a separate fat JAR for the webserver.

Also, as for that arbitrary attachment uploading; I had to keep that to keep the IRS sample working (Oracle rate upload). If there is a way to arbitrarily extend the RPC interface I can do that in the IRS sample, otherwise I'd like to keep that change to another PR to avoid making this one even larger.

Before I continue we need to agree on the scope of this particular PR and then the extension work.

mikehearn commented:
OK, fair enough then, if splitting it out into a totally separate app (like just a WAR that we point a jetty at) makes the PR too large, we can revisit in a second iteration. I'll continue the review.

Clintonio commented:
@mikehearn thanks. It's split at the point where it really matters for testing (process level). The next steps will take significantly less time.

Clintonio commented on core/src/main/kotlin/net/corda/core/node/services/Services.kt:
```diff
@@ -216,6 +218,16 @@ interface KeyManagementService {
     fun freshKey(): KeyPair
 }
 
+// TODO: Move and document
```
Documented, not moved yet.

Clintonio commented on node/src/integration-test/kotlin/net/corda/node/driver/DriverTests.kt:
```diff
@@ -1,5 +1,6 @@
 package net.corda.node.driver
 
+import com.google.common.net.HostAndPort
```
Really? It's the one used in other places I've checked. What's the correct one?

exFalso replied:
oops my bad sry

mikehearn commented on node/src/main/kotlin/net/corda/node/utilities/ConfigUtils.kt:
```diff
@@ -0,0 +1,9 @@
+import java.nio.file.Path
+import java.nio.file.Paths
+
+fun Config.getHostAndPort(name: String): HostAndPort = HostAndPort.fromString(getString(name))
```
Maybe just stuff these in Utilities? Seems like a new file for 2 trivial functions is a bit overkill

Clintonio replied:
That will require pulling typesafe config, an implementation detail of node, into core. The `utilities` folder in Node already has a fair number of utils files. We could create another uber `NodeUtils.kt` or accept some overhead in terms of file count.

mikehearn replied:
Yeah just add a `// TODO Create a NodeUtilities file to gather such things in one place` comment and we can revisit it later.

mikehearn commented on samples/irs-demo/src/main/kotlin/net/corda/irs/api/NodeInterestRates.kt:
```diff
@@ -110,7 +111,10 @@ object NodeInterestRates {
             val fixes = parseFile(data.bufferedReader().readText())
-            oracle.knownFixes = fixes
+            // TODO: Look into why knownFixes requires a transaction
+            transaction {
```
I think this is normally `databaseTransaction`?

Clintonio replied:
It looks like it'll need a database instance. After a brief look it doesn't seem like a trivial change unless I am missing something. I'll get back to this one after the reading and addressing the other comments.

Clintonio replied:
Fixed. Moved to the RPC call site.

Clintonio commented:
@mnesbit @mikehearn @exFalso comments addressed. Please follow up.

Clintonio commented:
I am going to squash this now. Lots of pointless commits and I want to start the next phase.

mikehearn commented:
Great. Let's make sure to suppress the deprecation warnings at the use sites so we stay warning-free.

Clintonio commented on node/src/main/kotlin/net/corda/node/Corda.kt:
```diff
@@ -78,33 +80,34 @@ fun main(args: Array<String>) {
     log.info("VM ${info.vmName} ${info.vmVendor} ${info.vmVersion}")
     log.info("Machine: ${InetAddress.getLocalHost().hostName}")
     log.info("Working Directory: ${cmdlineOptions.baseDirectory}")
+    log.info("Starting as webserver: ${cmdlineOptions.isWebserver}")
```
Fixed.

Clintonio commented on node/src/main/kotlin/net/corda/node/driver/Driver.kt:
```diff
@@ -318,25 +327,20 @@ open class DriverDSL(
         } catch(e: Exception) {
-            log.error("Could not query node info at $url due to an exception.", e)
-            return null
+            log.debug("Retrying query node info at $nodeAddress")
```
I've made them infinite on the Webserver calling the RPC, but making this infinite in the driver has issues with tests. I can do it, but it may have some unexpected pitfalls.

Clintonio commented on node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt:
```diff
@@ -228,6 +221,10 @@ abstract class AbstractNode(open val configuration: NodeConfiguration,
             customServices.addAll(buildPluginServices(tokenizableServices))
 
+            val uploaders: List<FileUploader> = listOf(storageServices.first.attachments as NodeAttachmentService) +
+                    customServices.filter { it is AcceptsFileUpload }.map { it as AcceptsFileUpload }
```
Thanks.

Clintonio commented on node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt:
```diff
@@ -104,6 +108,10 @@ class CordaRPCOpsImpl(
     override fun attachmentExists(id: SecureHash) = services.storageService.attachments.openAttachment(id) != null
     override fun uploadAttachment(jar: InputStream) = services.storageService.attachments.importAttachment(jar)
     override fun currentNodeTime(): Instant = Instant.now(services.clock)
+    override fun uploadFile(dataType: String, name: String?, file: InputStream): String {
```
That data type is not actually mime type. It's some sort of identifier for upload services to inspect to determine if they want to handle that particular file. File extensions can be handled by checking name, and MIME types, well, that can also be handled by the service when reading the input stream. 

I mostly copied this interface from the older data upload servlet _except_ I dropped the extension parameter because it can be handled inside the service.

Clintonio commented on gradle-plugins/cordformation/src/main/resources/net/corda/plugins/runnodes:
```diff
@@ -24,6 +24,16 @@ if which osascript >/dev/null; then
             first=false
         fi
     done
+    for dir in `ls`; do
```
The process will just exit instantly if it has no web configuration. I can add a more complex check on systems with bash, but for Windows I am not sure how easy it will be.

Clintonio replied:
The reason I left it this way is because I hope we'll be removing these soon with Chris' new system.

Clintonio commented:
I am going to do that squash now, I hope it doesn't lose too much.

@mikehearn ready for more review.

Clintonio commented on node/src/main/kotlin/net/corda/node/driver/Driver.kt:
```diff
@@ -485,13 +521,37 @@ open class DriverDSL(
-                    // the web api address to be bound as well, as that starts after the services. Needs rethinking.
-                    addressMustBeBound(executorService, nodeConf.webAddress)
-            ).map { process }
+            // TODO There is a race condition here. Even though the messaging address is bound it may be the case that
```
It does wait on the artemis port now, but it's as enough as the web was before. I think it's fine to wait on the port, the queryNodeInfo already waits on the RPC.

exFalso replied:
Fair enough, in that case the TODO can be removed

exFalso commented on node/src/main/kotlin/net/corda/node/driver/Driver.kt:
```diff
@@ -318,25 +327,20 @@ open class DriverDSL(
-            return om.readValue(conn.inputStream, NodeInfo::class.java)
+    private fun queryNodeInfo(nodeAddress: HostAndPort, sslConfig: SSLConfiguration): NodeInfo? {
+        var retries = 0
+        while (retries < 5) try {
```
Continuing the discussion (can't reply to original for some reason).

What I meant was we shouldn't have a cap on retries, it should just keep going. (I think the signature used to be nullable too, which is wrong). Same goes for the webserver polling

Clintonio replied:
The original version is rebased and overwritten I think. Okay, so infinite retries. We'll have to find a way to deal with issues with that on the CI server at one point. Hopefully though the problem of nodes not starting becomes a distant memory over time.

I'll go with infinite retries tomorrow.

exFalso replied:
Ping on this one

exFalso commented on node/src/main/kotlin/net/corda/node/internal/Node.kt:
```diff
@@ -305,43 +173,27 @@ class Node(override val configuration: FullNodeConfiguration,
+        thread(name = "WebServer") {
+            networkMapRegistrationFuture.getOrThrow()
+            // Begin exporting our own metrics via JMX.
+            JmxReporter.
```
Have we figured out what to do with JMX?

Clintonio replied:
In my original comment I plan to leave that for a bit later. I don't want to go down a hole of learning about JMX and sending it via RPC or some other mechanism (or just another port) right now. But we absolutely do need to get it back in a following PR.

mikehearn replied:
Jolokia can use a little embedded web server to export just that stuff, but JMX is often used in combination with an agent anyway.

Clintonio commented on docs/source/setting-up-a-corda-network.rst:
```diff
@@ -58,7 +59,9 @@ You should see a banner, some log lines and eventually ``Node started up and reg
 
-In terms of process management there is no prescribed method. You may start the jars by hand or perhaps use systemd and friends.
+In terms of process management there is no prescribed method. You may start the jars by hand or perhaps use systemd and friends. If you have
+deployed your nodes via `Cordformation`_ you can run them with the ``runnodes`` scripts in ``build/nodes`` of your project. This is primarily
```
Ah yes, I had written a headless noderunner when I wrote that. Maybe I can get that back in. I'll think about it.

Clintonio replied:
Reverted.

Clintonio commented:
@exFalso @mikehearn ready for re-review.

exFalso commented on docs/source/corda-plugins.rst:
```diff
@@ -34,21 +34,16 @@ of the node internal subsystems.
-    initiate threads. (N.B. The intent is to move the Web support into a 
-    separate helper process using the RPC mechanism to control access.) 
+    classes. These classes will be constructed by the bundled web server
+    and must have a single argument constructor taking a ``CordaRpcOps``
```
typo `CordaRpcOps` -> `CordaRPCOps

Clintonio replied:
Fixed.

exFalso commented on node/src/main/kotlin/net/corda/node/Corda.kt:
```diff
@@ -78,33 +80,38 @@ fun main(args: Array<String>) {
-        val node = conf.createNode()
-        node.start()
-        printPluginsAndServices(node)
+        if(!cmdlineOptions.isWebserver) {
```
whitespace 🚓

Clintonio replied:
Fixed

exFalso commented on node/src/main/kotlin/net/corda/node/webserver/WebServer.kt:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    private fun retryConnectLocalRpc(): CordaRPCOps {
+        while(true) {
```
whitespace 🚓

exFalso commented on samples/bank-of-corda-demo/src/integration-test/kotlin/net/corda/bank/BankOfCordaRPCClientTest.kt:
```diff
@@ -14,6 +14,7 @@ import net.corda.node.services.transactions.SimpleNotaryService
 import net.corda.testing.BOC_PARTY_REF
 import net.corda.testing.expect
 import net.corda.testing.expectEvents
+import net.corda.node.utilities.getHostAndPort
```
This isn't needed anymore

Clintonio replied:
Fixed.

exFalso commented on samples/irs-demo/src/main/kotlin/net/corda/irs/api/NodeInterestRates.kt:
```diff
@@ -17,12 +17,10 @@ import net.corda.core.utilities.ProgressTracker
+import net.corda.node.utilities.*
 import org.jetbrains.exposed.sql.ResultRow
 import org.jetbrains.exposed.sql.statements.InsertStatement
+import org.jetbrains.exposed.sql.transactions.transaction
```
Not needed

exFalso commented on samples/irs-demo/src/main/kotlin/net/corda/simulation/IRSSimulation.kt:
```diff
@@ -11,13 +11,15 @@ import net.corda.core.contracts.UniqueIdentifier
 import net.corda.core.flatMap
 import net.corda.core.flows.FlowStateMachine
 import net.corda.core.map
+import net.corda.core.messaging.CordaRPCOps
```
These imports not needed

Clintonio replied:
Fixed.

exFalso commented on samples/trader-demo/src/integration-test/kotlin/net/corda/traderdemo/TraderDemoTest.kt:
```diff
@@ -8,6 +8,7 @@ import net.corda.node.driver.driver
 import net.corda.node.services.User
 import net.corda.node.services.startFlowPermission
 import net.corda.node.services.transactions.SimpleNotaryService
+import net.corda.node.utilities.getHostAndPort
```
Not needed

Clintonio replied:
Fixed.

Clintonio commented:
@exFalso Somehow I can't reply to some of your comments.

Clintonio commented:
@exFalso I have addressed your mysterious comments.

exFalso commented:
Looking good, my only remaining concern is with `Party` parsing including an RPC call - can we look into why we need it to be done this way? I have a feeling we don't need this at all.

mikehearn commented:
I've approved this but please do add TODO comments in the various areas where the split still needs work. In particular:

* JSON support should become a part of the client lib, as the node won't care about JSON after the stuff is moved out.
* JMX export can be brought back but it doesn't need anything as big as a full jetty. The JRE has a tiny web server built in which I have a feeling Jolokia can use, but it should also be something you opt in to rather than being enabled by default. JMX sometimes supports changing the server as well as watching it, so it's a security sensitive thing to have.
* Demos should eventually become WARs that can be dropped into a normal app container, or whatever it is the cool kids do these days :) But basically the remaining web server support should be moved out into the client library so any existing Java web app can start using the Corda node.
* Demo web server should show direct connect to the H2 database doing SQL queries, rather than doing queries via RPC.

Add these as TODOs to the right places and we're done for this sprint. 👍

Clintonio commented:
@exFalso I aimed to not change behaviour I didn't know the purpose of in this PR. I can shove a TODO on it for someone to look into?

Clintonio commented:
@mikehearn Will do. Though I'd like to discuss that last bullet on Slack at some point.
